x-pack/plugins/security/server/session_management/session.ts

```diff
@@ -0,0 +1,450 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import nodeCrypto, { Crypto } from '@elastic/node-crypto';
+import { promisify } from 'util';
+import { randomBytes, createHash } from 'crypto';
+import { Duration } from 'moment';
+import { KibanaRequest, Logger } from '../../../../../src/core/server';
+import { AuthenticationProvider } from '../../common/types';
+import { SecurityAuditLogger } from '../audit';
+import { ConfigType } from '../config';
+import { SessionIndex, SessionIndexValue } from './session_index';
+import { SessionCookie } from './session_cookie';
+
+/**
+ * The shape of the value that represents user's session information.
+ */
+export interface SessionValue {
+  /**
+   * Unique session ID.
+   */
+  sid: string;
+
+  /**
+   * Username this session belongs. It's defined only if session is authenticated, otherwise session
+   * is considered unauthenticated (e.g. intermediate session used during SSO handshake).
+   */
+  username?: string;
+
+  /**
+   * Name and type of the provider this session belongs to.
+   */
+  provider: AuthenticationProvider;
+
+  /**
+   * The Unix time in ms when the session should be considered expired. If `null`, session will stay
+   * active until the browser is closed.
+   */
+  idleTimeoutExpiration: number | null;
+
+  /**
+   * The Unix time in ms which is the max total lifespan of the session. If `null`, session expire
+   * time can be extended indefinitely.
+   */
+  lifespanExpiration: number | null;
+
+  /**
+   * Kibana server base path the session was created for.
+   */
+  path: string;
+
+  /**
+   * Session value that is fed to the authentication provider. The shape is unknown upfront and
+   * entirely determined by the authentication provider that owns the current session.
+   */
+  state: unknown;
+
+  /**
+   * Indicates whether user acknowledged access agreement or not.
+   */
+  accessAgreementAcknowledged?: boolean;
+
+  /**
+   * Additional information about the session value.
+   */
+  metadata: { index: SessionIndexValue };
```

Comment:
__note:__ I keep this value to use `index` operation for updates and not `_update` since we retrieve session from the index anyway and there is no reason to do that again (AFAIK `_update` will fetch full document before indexing anyway) + we need to know `primaryTerm`  and `sequenceNumber`.